torchtune/utils/collate.py

```diff
@@ -69,3 +71,73 @@ def padded_collate(
             value=padding_idx,
         )
     return {"tokens": input_ids.long(), "labels": labels.long()}
+
+
+def padded_collate_packed(
+    batch: List[PACK_TYPE],
+    device: torch.device = torch.device("cuda"),
+) -> Dict[str, torch.Tensor]:
+    """Collate packed sequences into a batch. Only convert the seq lens into
+    a block mask for use with attention. Tokens, labels, and input_pos are
+    already padded to the same length within :class:`~torchtune.datasets.PackedDataset`.
```

Comment:
Yeah I am also confused by the padding logic, but have a slightly different question. Iiuc:

(1) we pad each pack to `max_seq_len` in the packed dataset. We also append an element to the pack's seq_len field telling us how many padding tokens there are
(2) in the collator we then pad the seq_len field for each pack in the batch to be the same length
(3) we construct the mask using these padded seq_len fieldsing 

Is there something in (3) that necessitates the padding done in (2)? Maybe I'm being dense but I don't see it.. we either use flex attention, in which case we call `_get_document_ids_from_seq_lens`, or we don't, in which case we call `create_block_causal_mask`. In both cases we are just iterating over `seq_lens[batch_idx]`, right? So if I've already padded everything to a fixed length inside of the packed dataset, what is the value of also padding the seq lengths? 

Concretely, if I have pack 0 with (unpadded) sequence lengths [1, 3, 5] and pack 1 with (unpadded) sequence lengths [1, 2, 3, 4, 5] and my max_seq_len is 20, then in the packed dataset I will pad each pack and have new seq lengths [1, 3, 5, 11] and [1, 2, 3, 4, 5, 5], respectively. But then iiuc the collator will also pad [1, 3, 5, 11] -> [1, 3, 5, 11, 0, 0]. Why?

Reply:
Also @felipemello1's comment on moving all the padding into collate is interesting. I'm kinda inclined to agree; the one drawback could be that we lose that nice fixed-sequence-length property of our packed datasets that makes compile's life easier (though it's done fine with our unpacked variable-length sequences so maybe nbd). I do wonder if we should decouple those changes from this PR since it involves more nontrivial changes to the packed dataset (unless ofc it winds up being straightforward to do on top of the other stuff here)

Reply:
If we want to keep this property, we can still have it in the collator

Reply:
cc @gau-nernst mentioned that keep seq len fixed in PackedDataset might still be more efficient for compile even if it does compile successfully

Reply:
> Concretely, if I have pack 0 with (unpadded) sequence lengths [1, 3, 5] and pack 1 with (unpadded) sequence lengths [1, 2, 3, 4, 5] and my max_seq_len is 20, then in the packed dataset I will pad each pack and have new seq lengths [1, 3, 5, 11] and [1, 2, 3, 4, 5, 5], respectively. But then iiuc the collator will also pad [1, 3, 5, 11] -> [1, 3, 5, 11, 0, 0]. Why?

@ebsmothers I think your assessment is correct. We shouldn't need to pad seq lens as long as they each sum to max seq len